File: ExcludedSamples/Ribbon/VB/ControlHost/ControlHostLib/NumericUpDownHost.vb
Imports System
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Xml
Imports C1.Win.C1Ribbon

''' <summary>
''' The Ribbon element that contains a NumericUpDown.
''' </summary>
Public Class NumericUpDownHost
    Inherits RibbonControlHost

    Public Const XmlNodeName As String = "numericUpDownHost"

    ''' <summary>
    ''' Creates the NumericUpDown control and its host element.
    ''' </summary>
    Public Sub New()
        MyBase.New(New NumericUpDown())
    End Sub

    ''' <summary>
    ''' <inheritdoc />
    ''' </summary>
    Public Overrides ReadOnly Property ControlHostNodeName As String
        Get
            Return XmlNodeName
        End Get
    End Property

    ''' <summary>
    ''' Gets the hosted NumericUpDown control.
    ''' </summary>
    <Browsable(False)> _
    <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
    Public ReadOnly Property NumericUpDown() As NumericUpDown
        Get
            Return CType(Control, NumericUpDown)
        End Get
    End Property

    ''' <summary>
    ''' Gets or sets the value assigned to the NumericUpDown.
    ''' </summary>
    <Category("Data")> _
    <Description("Gets or sets the value assigned to the NumericUpDown.")> _
    Public Property Value() As System.Decimal
        Get
            Return NumericUpDown.Value
        End Get
        Set(ByVal value As System.Decimal)
            NumericUpDown.Value = value
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the minimum value for the NumericUpDown.
    ''' </summary>
    <Category("Behavior")> _
    <Description("Gets or sets the minimum value for the NumericUpDown.")> _
    <DefaultValue(0)> _
    Public Property Minimum() As System.Decimal
        Get
            Return NumericUpDown.Minimum
        End Get
        Set(ByVal value As System.Decimal)
            NumericUpDown.Minimum = value
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the maximum value for the NumericUpDown.
    ''' </summary>
    <Category("Behavior")> _
    <Description("Gets or sets the maximum value for the NumericUpDown.")> _
    <DefaultValue(100)> _
    Public Property Maximum() As System.Decimal
        Get
            Return NumericUpDown.Maximum
        End Get
        Set(ByVal value As System.Decimal)
            NumericUpDown.Maximum = value
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the value to increment or decrement the NumericUpDown when the up or down buttons are clicked.
    ''' </summary>
    <Category("Behavior")> _
    <Description("Gets or sets the value to increment or decrement the NumericUpDown when the up or down buttons are clicked.")> _
    <DefaultValue(1)> _
    Public Property Increment() As System.Decimal
        Get
            Return NumericUpDown.Increment
        End Get
        Set(ByVal value As System.Decimal)
            NumericUpDown.Increment = value
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the text that appears on the NumericUpDown.
    ''' </summary>
    <Browsable(False)> _
    <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
    Public Overrides Property Text() As String
        Get
            Return MyBase.Text
        End Get
        Set(ByVal value As String)
            MyBase.Text = value
        End Set
    End Property

    ''' <summary>
    ''' Occurs when the <see cref="Value"/> property changes.
    ''' </summary>
    <Category("Property Changed")> _
    <Description("Occurs when the Value property changes.")> _
    Public Event ValueChanged As EventHandler
    ''' <summary>
    ''' Raises the <see cref="ValueChanged"/> event.
    ''' </summary>
    Protected Overridable Sub OnValueChanged(ByVal e As EventArgs)
        RaiseEvent ValueChanged(Me, e)
    End Sub

    Protected Overrides Sub OnSubscribeControlEvents(ByVal ctrl As Control)
        MyBase.OnSubscribeControlEvents(ctrl)
        AddHandler CType(ctrl, NumericUpDown).ValueChanged, AddressOf HandleValueChanged
    End Sub
    Protected Overrides Sub OnUnsubscribeControlEvents(ByVal ctrl As Control)
        MyBase.OnUnsubscribeControlEvents(ctrl)
        RemoveHandler CType(ctrl, NumericUpDown).ValueChanged, AddressOf HandleValueChanged
    End Sub

    Private Sub HandleValueChanged(ByVal sender As Object, ByVal e As EventArgs)
        Me.OnValueChanged(e)
    End Sub

    Protected Overrides Sub LoadControlHostProperties(ByVal node As XmlNode)
        MyBase.LoadControlHostProperties(node)
        Value = Read(node, "value", 0)
        Minimum = Read(node, "minimum", 0)
        Maximum = Read(node, "maximum", 100)
        Increment = Read(node, "increment", 1)
    End Sub

    Protected Overrides Sub SaveControlHostProperties(ByVal writer As XmlWriter)
        MyBase.SaveControlHostProperties(writer)
        Write(writer, "value", Value, 0)
        Write(writer, "minimum", Minimum, 0)
        Write(writer, "maximum", Maximum, 100)
        Write(writer, "increment", Increment, 1)
    End Sub

    Private Shared Function Read(ByVal node As XmlNode, ByVal attName As String, ByVal defVal As Decimal) As Decimal
        Dim att As XmlAttribute = node.Attributes(attName)
        If att IsNot Nothing Then
            Return Decimal.Parse(att.Value)
        End If

        Return defVal
    End Function

    Private Shared Sub Write(ByVal writer As XmlWriter, ByVal attName As String, ByVal value As Decimal, ByVal defVal As Decimal)
        If value <> defVal Then
            writer.WriteAttributeString(attName, value.ToString())
        End If
    End Sub

End Class

File: ExcludedSamples/NetFramework/Ribbon/CS/ControlHost/ControlHostLib/NumericUpDownHost.cs
using System;
using System.ComponentModel;
using System.Windows.Forms;
using System.Xml;

using C1.Win.C1Ribbon;

namespace ControlHostLib
{
    /// <summary>
    /// The Ribbon element that contains a NumericUpDown.
    /// </summary>
    public class NumericUpDownHost : RibbonControlHost
    {
        public const string XmlNodeName = "numericUpDownHost";

        //===========================================================================
        #region ** ctor

        /// <summary>
        /// Creates the NumericUpDown control and its host element.
        /// </summary>
        public NumericUpDownHost()
            : base(new NumericUpDown())
        {
        }

        #endregion

        //===========================================================================
        #region ** properties

        /// <inheritdoc />
        public override string ControlHostNodeName
        {
            get { return XmlNodeName; }
        }

        /// <summary>
        /// Gets the hosted NumericUpDown control.
        /// </summary>
        [Browsable(false)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public NumericUpDown NumericUpDown
        {
            get { return (NumericUpDown)Control; }
        }

        /// <summary>
        /// Gets or sets the value assigned to the NumericUpDown.
        /// </summary>
        [Category("Data")]
        [Description("Gets or sets the value assigned to the NumericUpDown.")]
        [DefaultValue(0)]
        public decimal Value
        {
            get { return NumericUpDown.Value; }
            set { NumericUpDown.Value = value; }
        }

        /// <summary>
        /// Gets or sets the minimum value for the NumericUpDown.
        /// </summary>
        [Category("Behavior")]
        [Description("Gets or sets the minimum value for the NumericUpDown.")]
        [DefaultValue(0)]
        public decimal Minimum
        {
            get { return NumericUpDown.Minimum; }
            set { NumericUpDown.Minimum = value; }
        }

        /// <summary>
        /// Gets or sets the maximum value for the NumericUpDown.
        /// </summary>
        [Category("Behavior")]
        [Description("Gets or sets the maximum value for the NumericUpDown.")]
        [DefaultValue(100)]
        public decimal Maximum
        {
            get { return NumericUpDown.Maximum; }
            set { NumericUpDown.Maximum = value; }
        }

        /// <summary>
        /// Gets or sets the value to increment or decrement the NumericUpDown when the up or down buttons are clicked.
        /// </summary>
        [Category("Behavior")]
        [Description("Gets or sets the value to increment or decrement the NumericUpDown when the up or down buttons are clicked.")]
        [DefaultValue(1)]
        public decimal Increment
        {
            get { return NumericUpDown.Increment; }
            set { NumericUpDown.Increment = value; }
        }

        /// <summary>
        /// Gets or sets the text that appears on the NumericUpDown.
        /// </summary>
        [Browsable(false)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public override string Text
        {
            get { return string.Empty; }
            set { }
        }

        #endregion

        //===========================================================================
        #region ** events

        /// <summary>
        /// Occurs when the <see cref="Value"/> property changes.
        /// </summary>
        [Category("Property Changed")]
        [Description("Occurs when the Value property changes.")]
        public event EventHandler ValueChanged;
        /// <summary>
        /// Raises the <see cref="ValueChanged"/> event.
        /// </summary>
        protected virtual void OnValueChanged(EventArgs e)
        {
            if (ValueChanged != null)
                ValueChanged(this, e);
        }

        #endregion

        //===========================================================================
        #region ** methods

        protected override void OnSubscribeControlEvents(Control control)
        {
            base.OnSubscribeControlEvents(control);
            ((NumericUpDown)control).ValueChanged += new EventHandler(HandleValueChanged);
        }
        protected override void OnUnsubscribeControlEvents(Control control)
        {
            base.OnUnsubscribeControlEvents(control);
            ((NumericUpDown)control).ValueChanged -= new EventHandler(HandleValueChanged);
        }

        void HandleValueChanged(object sender, EventArgs e)
        {
            this.OnValueChanged(e);
        }

        protected override void LoadControlHostProperties(XmlNode node)
        {
            base.LoadControlHostProperties(node);

            Value = Read(node, "value", 0);
            Minimum = Read(node, "minimum", 0);
            Maximum = Read(node, "maximum", 100);
            Increment = Read(node, "increment", 1);
        }

        protected override void SaveControlHostProperties(XmlWriter writer)
        {
            base.SaveControlHostProperties(writer);

            Write(writer, "value", Value, 0);
            Write(writer, "minimum", Minimum, 0);
            Write(writer, "maximum", Maximum, 100);
            Write(writer, "increment", Increment, 1);
        }

        static decimal Read(XmlNode node, string attName, decimal defVal)
        {
            XmlAttribute att = node.Attributes[attName];
            if (att != null)
            {
                return Decimal.Parse(att.Value);
            }
            return defVal;
        }

        static void Write(XmlWriter writer, string attName, decimal value, decimal defVal)
        {
            if (value != defVal)
            {
                writer.WriteAttributeString(attName, value.ToString());
            }
        }

        #endregion
    }
}
